src/V3SignalStrength.cpp

```diff
@@ -0,0 +1,145 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Deals with signal strength
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2022 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Global.h"
+#include "V3SignalStrength.h"
+#include "V3Ast.h"
+#include "V3Stats.h"
+#include "V3Inst.h"
+#include "V3Graph.h"
+
+#include <algorithm>
+#include <map>
+
+//######################################################################
+
+class SignalStrengthVisitor final : public VNVisitor {
+
+    // TYPES
+    using Assigns = std::vector<AstAssignW*>;
+    using VarToAssignsMap = std::unordered_map<AstVar*, Assigns>;
+
+    // MEMBERS
+    VarToAssignsMap m_assigns;  // Assigns in current module
+
+    // VISITORS
+    virtual void visit(AstAssignW* nodep) {
+        if (AstVarRef* varRefp = VN_CAST(nodep->lhsp(), VarRef))
+            m_assigns[varRefp->varp()].push_back(nodep);
+    }
+
+    AstAssign* getStrengthAssignmentp(FileLine* fl, AstVar* strengthVarp, int strengthLevel,
+                                      AstNode* assignedValuep, int compareConstp) {
+        return new AstAssign(
+            fl, new AstVarRef(fl, strengthVarp, VAccess::WRITE),
+            new AstCond(fl,
+                        new AstLogAnd(fl,
+                                      new AstLt(fl, new AstVarRef(fl, strengthVarp, VAccess::READ),
+                                                new AstConst(fl, strengthLevel)),
+                                      new AstEqCase(fl, assignedValuep,
+                                                    new AstConst(fl, AstConst::WidthedValue(), 1,
+                                                                 compareConstp))),
+                        new AstConst(fl, strengthLevel),
+                        new AstVarRef(fl, strengthVarp, VAccess::READ)));
+    }
+
+    virtual void visit(AstNodeModule* nodep) override {
+        UINFO(8, nodep << endl);
+        iterateChildren(nodep);
+        for (auto& varpAssigns : m_assigns) {
```

Comment:
Do we need to store this in a map? I think this can be done in `visit(AstAssignW*)`, while iterating the assigns. Utilize the `user{1,2,3,4,5}` pointers.

Reply:
But there can be more than 5 nets in one module.

Reply:
That's why I said this can be done in `visit(AstAssignW*)` (that inner loop at least), not in `visit(AstNodeModule*)`. The new vars and the conditional `AssignW` can be created in `V3SignalStrength.cpp:76`. Put the shared info in `user*`. Let's not re-iterate over nodes if we can avoid it. See how other visitors utilize this, e.g. `V3Delayed` puts objects there for future re-use.

I may be wrong though, and this is needed.

Reply:
I know what you mean now. This would require to store `__s0`, `__s1` variables, their scope (or we can use just one for computing strengths for all of variables) and first assignment to that variable (to remove it when we encounter second assignment), so we would need to use 3 `user*p` pointers.

We know that these `__s*` variables will be needed at the same moment when we should remove the first assignment, so it is possible to use the same `user*p` pointer for assignment and for one of the variables, but it will make the situation that meaning of that pointer will change during handling of 2nd assignment. So let's say that 2 using of 2 userp pointers is the minimum.

As we previously talked, it may be required to move content of this visitor to some already existing visitor. Using of `user*p` pointers will make it more complicated. On the other hand, as I checked, V3Width is using none of these user*p, so maybe we don't need to care about it, if I have to move this to exactly V3Width visitor.

In `visit(AstNodeModule*)` we handle for 2nd time only these assignment that require any work (they have signal strengths or are one of multiple assignments to the same variable). I don't know how often it happens, but handling of them involve creating additional expressions, which is also a cost.

So I think it is possible to use `user*p`, but I think that benefits from it will be small and it will complicate the code. What do you think?

Reply:
Ok, let's leave it up to Verilator maintainers.